src/Apropos/Gen/BacktrackingTraversal.hs

```diff
@@ -0,0 +1,125 @@
+module Apropos.Gen.BacktrackingTraversal (
+  Traversal (..),
+  traversalAsGen,
+  traversalContainRetry,
+) where
+
+import Apropos.Gen
+import Apropos.HasPermutationGenerator.Morphism
+import Control.Monad ((>=>))
+import Control.Monad.Trans.Except (runExceptT)
+import Control.Monad.Trans.Writer (runWriterT)
+import Data.String (fromString)
+import Hedgehog (PropertyT)
+import Hedgehog qualified as H
+
+data Traversal p m where
+  Source :: Gen m -> Traversal p m
+  Traversal :: Traversal p m -> (m -> Gen [Morphism p m]) -> Traversal p m
+
+traversalAsGen :: Traversal p m -> Gen m
+traversalAsGen (Source g) = g
+traversalAsGen (Traversal m g) = do
+  n <- traversalAsGen m
+  t <- g n
+  foldl (>=>) pure (morphism <$> t) n
+
+traversalContainRetry :: Show m => Int -> Traversal p m -> PropertyT IO m
+traversalContainRetry retries m = do
+  r <- traversalWithRetries retries m
+  case r of
+    Nothing -> H.footnote "retry limit reached" >> H.discard
+    Just so -> pure so
+
+traversalWithRetries :: forall p a. Show a => Int -> Traversal p a -> PropertyT IO (Maybe a)
+traversalWithRetries retries m =
+  backtrackingRetryTraversals retries (forAllRetryToMaybeScale s) (genTraversal <$> t)
+  where
+    st :: (Gen a, [a -> Gen [Morphism p a]])
+    st = unTraversal m
+    s :: Gen a
+    s = fst st
+    t :: [a -> Gen [Morphism p a]]
+    t = snd st
+    genTraversal :: (a -> Gen [Morphism p a]) -> (a -> PropertyT IO [Int -> a -> PropertyT IO (Maybe a)])
+    genTraversal gt = \a -> do
+      tr <- forAllWith (unwords . (name <$>)) $ prune (gt a)
```

Comment:
We use the name of the morphism here - otherwise it is just being treated as a `m -> Gen m`. This will let us see what went wrong if for some reason generation fails when trying to build a path.